MSC4391: Simplified in-room bot commands

Gnuxie opened the pull request:
[Rendered](https://github.com/Gnuxie/matrix-doc/blob/gnuxie/simplified-in-room-bot-commands/proposals/4391-simplified-in-room-bot-commands.md)

Signed-off-by: Gnuxie [Gnuxie@protonmail.com](mailto:Gnuxie@protonmail.com)

---

This MSC incorporates work from [MSC4332](https://github.com/matrix-org/matrix-spec-proposals/pull/4332)

turt2live committed:
MSC: In-room bot commands

Start off by incorporating MSC4332.

Gnuxie committed:
Writeup key semantics

Gnuxie committed:
JSON-RPC similarity acknowledgement

Gnuxie committed:
Use multiple state events for command descriptions.

Gnuxie committed:
Union and array types.

Gnuxie committed:
Literal types.

Gnuxie committed:
Resolve enabled commands TODO.

Not really needed when the command description state event is scoped
by sender.

Gnuxie committed:
More schema types.

Gnuxie committed:
Profile of command description sender.

Gnuxie committed:
Write new background to the MSC and the problems we want to solve.

Even though the MSC is simpler, we're more ambitious and able to do
more with it.

Gnuxie committed:
MSC4391

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
Implementation requirements:

* [ ] Client
* [ ] Bot

tulir replied:
https://github.com/gomuks/gomuks/pull/675 is an implementation of client and technically also bot, converted from MSC4332. It assumes most of my suggestions in other comments are applied

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+  // derived from `sha256(designator.join('') + mxid)`
+  "state_key": "JBDLR6YMe+72yqsEMi/MVdTmjN3ynPThMz+M7QLATZQ=",
+  "content": {
+    "designator": ["ban"],
```
Why is this an array? If it's for subcommands, why not just have a space-separated string? Also, aliases are important

Gnuxie replied:
Yeah we should make this a single string that may be space separated.

The difficulty with aliases is whether a room admin is supposed to be able to set those or just the bot. Because one of the security invariants is that the `sender` of the command description describes which bot it is for, and this is in order to ensure bots don't sabotage each other or something.

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+  [event ID](https://spec.matrix.org/v1.15/appendices/#event-ids).
+
+When an object is provided as the type schema, the type is a derived from a
+schema described by the property `schema_type`:
```
`type` being either a string or an object is annoying to parse in some languages. It would be preferable if it was always an object with varying keys. e.g. `{"schema_type": "primitive", "type": "room_id"}`

Gnuxie replied:
Should the type property also just be changed to `schema` then on the parameter description?

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+- The `union` schema type specifies the types of the variants with the
+  `variants` property. Which is an array of type schema.
+- The `literal` schema type specifies a literal value with the `value` property
+  and the type of the literal value with the `literal_type` property.
```
I don't think there's any reason to specify a reference type here, so is `literal_type` needed if it can only be a string, integer or boolean? Could just be the `value` field that contains one of the three types (never an object/array/null)

tulir replied:
Room and event IDs have multiple textual representations, but those might be fine-ish. I don't see any reason to allow arrays though. Parsing those would be even more difficult, plus they'd cause more schema nesting issues (union of literals is allowed, but an union of literals of arrays would be bad)

Lack of arrays won't prevent anything, you can always have a literal comma-separated string or whatever. At worst it'll have slightly worse UI, but not sure if literal arrays would ever have good UI either. You can't put an array into a `<select>` UI without stringifying it first, at which point the command definition could've just stringified it ahead of time to save everyone the trouble.

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+- The `array` schema type specifies the type of the items with the `items`
+  property.
+- The `union` schema type specifies the types of the variants with the
+  `variants` property. Which is an array of type schema.
```
Also, should arrays be allowed only as the final parameter (i.e. as a variadic parameter)? Multiple arrays or other parameters after arrays could get complicated

tulir commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+  "type": "m.bot.command_description",
+  "sender": "@draupnir:draupnir.space",
+  // derived from `sha256(designator.join('') + mxid)`
+  "state_key": "JBDLR6YMe+72yqsEMi/MVdTmjN3ynPThMz+M7QLATZQ=",
```
Is the state key format mandatory or just a suggestion? If it's mandatory, clients could validate it as a form of deduplication. If not, clients will probably need to ignore state keys and implement some deduplication on the command itself

tulir replied:
MUST would be nice since that allows clients to discard any duplicate commands simply by enforcing the state key format, but just needs some keyword. If it's not MUST, clients will ignore the state key entirely and implement deduplication elsewhere (deduplication of some kind is probably necessary, duplicate command definitions don't make much sense)

Gnuxie committed:
designator -> command.

https://github.com/matrix-org/matrix-spec-proposals/pull/4391#discussion_r2646280698

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Sounder reasoning about fallback.

https://github.com/matrix-org/matrix-spec-proposals/pull/4391#discussion_r2646197140

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Rename parameter description "type" prop to "schema".

Enable ease of parsing
https://github.com/matrix-org/matrix-spec-proposals/pull/4391#discussion_r2646287889

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Rename required -> optional.

https://github.com/matrix-org/matrix-spec-proposals/pull/4391#discussion_r2646313869

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,332 @@
+          "m.text": [{ "body": "Whether to apply this to the policy" }]
+        },
+        // This argument is not required
+        "required": false
```
I didn't include defaults at this stage because they can also be provided by partial commands and interactive prompts ie the way suggestions work in draupnir. I guess it might be ok if it's treated like a literal default.

Gnuxie committed:
Restrict unions and arrays.

https://github.com/matrix-org/matrix-spec-proposals/pull/4391#discussion_r2646300409

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Resolve array schema in the middle.

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Deduplicating command descriptions explicitly.

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Restrict literals.

Co-authored-by: Tulir Asokan <tulir@maunium.net>

Gnuxie committed:
Note about schema validation being optional for clients.

Gnuxie commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,374 @@
+- Using state events limits a bot's ability to advertise commands if it isn't
+  given power to do so.
+
+## Alternatives
```
I'm in the process of adding a "response" mixin in a follow up MSC. It would be really useful to also be able to proxy the commands into a mutual room (like draupnir's command room). See https://github.com/the-draupnir-project/Draupnir/issues/1041#issuecomment-4071332194

TheArcaneBrony commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,374 @@
+  "state_key": "JBDLR6YMe+72yqsEMi/MVdTmjN3ynPThMz+M7QLATZQ=",
+  "content": {
+    // space separated string for commands such as "rooms add"
+    "command": "ban",
```
Is there a way to define command aliases?

TheArcaneBrony commented on proposals/4391-simplified-in-room-bot-commands.md:
```diff
@@ -0,0 +1,374 @@
+  "content": {
+    // space separated string for commands such as "rooms add"
+    "command": "ban",
+    "parameters": [
```
this seems to be missing support for positional parameters?
